File: BillPayments/Core/InMemoryKeyValueStore.swift
final class InMemoryKeyValueStore: KeyValueStore {
    var failWrites = false
    private(set) var values: [String: String] = [:]

    func read(_ key: String) -> String? { values[key] }

    func write(_ key: String, _ value: String) async throws {
        if failWrites { throw StorageError() }
        values[key] = value
    }

    func remove(_ key: String) async throws {
        values[key] = nil
    }
}

File: app/src/main/java/com/billpayments/core/storage/InMemoryKeyValueStore.kt
package com.billpayments.core.storage

class InMemoryKeyValueStore(var failWrites: Boolean = false) : KeyValueStore {
    private val values = mutableMapOf<String, String>()

    override fun read(key: String): String? = values[key]

    override suspend fun write(key: String, value: String) {
        if (failWrites) throw StorageException()
        values[key] = value
    }

    override suspend fun remove(key: String) {
        values.remove(key)
    }
}
